internal/core/operations/switch_active_version/switch_active_version_executor.go

```diff
@@ -0,0 +1,233 @@
+		roomManager:        roomManager,
+		schedulerManager:   schedulerManager,
+		roomsBeingReplaced: &sync.Map{},
+		newCreatedRooms:    newCreatedRoomsMap,
+		locker:             sync.Mutex{},
+	}
+}
+
+// Execute the process of switching a scheduler active version consists of the following:
+// 1. Creates "replace" goroutines (same number as MaxSurge);
+// 2. Each goroutine will listen to a channel and create a new room using the
+//    new configuration. After the room is ready, it will then delete the room
+//    being replaced;
+// 3. List all game rooms that need to be replaced and produce them into the
+//    replace goroutines channel;
+// 4. Switch the active version
+func (ex *SwitchActiveVersionExecutor) Execute(ctx context.Context, op *operation.Operation, definition operations.Definition) error {
+	logger := zap.L().With(
+		zap.String("scheduler_name", op.SchedulerName),
+		zap.String("operation_definition", definition.Name()),
+		zap.String("operation_id", op.ID),
+	)
+	logger.Debug("start switching scheduler active version")
+
+	updateDefinition, ok := definition.(*SwitchActiveVersionDefinition)
+	if !ok {
+		return fmt.Errorf("the definition is invalid. Should be type SwitchActiveVersionDefinition")
+	}
+	scheduler := &updateDefinition.NewActiveScheduler
+
+	maxSurgeNum, err := ex.roomManager.SchedulerMaxSurge(ctx, scheduler)
+	if err != nil {
+		return fmt.Errorf("error fetching scheduler max surge: %w", err)
+	}
+
+	err = ex.startReplaceRoomsLoop(ctx, logger, maxSurgeNum, *scheduler)
+	if err != nil {
+		return err
+	}
+
+	err = ex.schedulerManager.SwitchActiveScheduler(ctx, scheduler)
+	if err != nil {
+		logger.Error("Error switching active scheduler version on scheduler manager")
+		return err
+	}
+
+	ex.clearNewCreatedRooms(op.SchedulerName)
+	logger.Debug("scheduler update finishes with success")
+	return nil
+}
+
+func (ex *SwitchActiveVersionExecutor) OnError(ctx context.Context, op *operation.Operation, definition operations.Definition, executeErr error) error {
+	logger := zap.L().With(
+		zap.String("scheduler_name", op.SchedulerName),
+		zap.String("operation_definition", definition.Name()),
+		zap.String("operation_id", op.ID),
+	)
+	logger.Info("starting OnError routine")
+
+	err := ex.deleteNewCreatedRooms(ctx, logger, op.SchedulerName)
+	ex.clearNewCreatedRooms(op.SchedulerName)
+	if err != nil {
+		return err
+	}
+
+	logger.Debug("finished OnError routine")
+	return nil
+}
+
+func (ex *SwitchActiveVersionExecutor) Name() string {
+	return OperationName
+}
+
+func (ex *SwitchActiveVersionExecutor) replaceRoom(logger *zap.Logger, wg *sync.WaitGroup, roomsChan chan *game_room.GameRoom, roomManager interfaces.RoomManager, scheduler entities.Scheduler) error {
+	defer wg.Done()
+
+	// we're going to use a separated context for each replaceRoom since we
+	// don't want to cancel the replace in the middle (like creating a room and
+	// then left the old one (without deleting it).
+	ctx := context.Background()
+
+	for {
```

Comment:
@caiordjesus I see value in logging if we have some error in creating or deleting rooms here